titiler/cmr/timeseries.py

```diff
@@ -0,0 +1,612 @@
+from datetime import datetime, timedelta
+from enum import Enum
+from types import DynamicClassAttribute
+from typing import Annotated, Any, Dict, List, Literal, Optional, Tuple, Union
+from urllib.parse import urlencode
+
+import earthaccess
+import httpx
+from attrs import define
+from dateutil.relativedelta import relativedelta
+from fastapi import Body, Depends, Path, Query, Request, Response
+from fastapi.exceptions import HTTPException
+from fastapi.responses import StreamingResponse
+from geojson_pydantic import Feature, FeatureCollection
+from geojson_pydantic.geometries import Geometry
+from PIL import Image
+from pydantic import BaseModel
+
+from titiler.cmr.dependencies import ConceptID
+from titiler.cmr.factory import Endpoints
+from titiler.core.algorithm import algorithms as available_algorithms
+from titiler.core.dependencies import CoordCRSParams, DefaultDependency, DstCRSParams
+from titiler.core.factory import FactoryExtension
+from titiler.core.models.mapbox import TileJSON
+from titiler.core.models.responses import Statistics
+from titiler.core.resources.enums import ImageType
+from titiler.core.resources.responses import GeoJSONResponse
+
+timeseries_img_endpoint_params: Dict[str, Any] = {
+    "responses": {
+        200: {
+            "content": {
+                "image/gif": {},
+            },
+            "description": "Return an image.",
+        }
+    },
+    "response_class": Response,
+}
+
+
+class TimeseriesMediaType(str, Enum):
+    """Responses Media types formerly known as MIME types."""
+
+    gif = "image/gif"
+
+
+class TimeseriesImageType(str, Enum):
+    """Available output image type."""
+
+    gif = "gif"
+
+    @DynamicClassAttribute
+    def mediatype(self):
+        """Return image media type."""
+        return TimeseriesMediaType[self._name_].value
+
+
+TimeseriesStatistics = Dict[str, Statistics]
+
+
+class TimeseriesStatisticsInGeoJSON(BaseModel):
+    """Statistics model in geojson response."""
+
+    statistics: TimeseriesStatistics
+
+    model_config = {"extra": "allow"}
+
+
+TimeseriesStatisticsGeoJSON = Union[
+    FeatureCollection[Feature[Geometry, TimeseriesStatisticsInGeoJSON]],
+    Feature[Geometry, TimeseriesStatisticsInGeoJSON],
+]
+
+TimeseriesTileJSON = Dict[str, TileJSON]
+
+
+@dataclass
+class TimeseriesParams(DefaultDependency):
+    """Timeseries parameters"""
+
+    start_datetime: Annotated[
+        Optional[str],
+        Query(
+            description="Start datetime for timeseries request",
+        ),
+    ] = None
+    end_datetime: Annotated[
+        Optional[str],
+        Query(
+            description="End datetime for timeseries request",
+        ),
+    ] = None
+    step: Annotated[
+        Optional[str],
+        Query(
+            description="Time step between timeseries intervals, expressed as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations)"
+        ),
+    ] = None
+    step_idx: Annotated[
```

Comment:
I thought `step_idx` would be useful in the case where you have a timeseries that you are interested in, but for some reason you only want to fetch results for a single time step at a time.

You could make requests like this:
```python
# format the base request url
base_request_url = "/timeseries/statistics?concept_id=1234&start_datetime=2024-10-01T00:00:01Z&end_datetime=2024-10-10T00:00:01Z"

first_request = f"{base_request_url}&step_idx=1"

second_request = f"{base_request_url}&step_idx=2"
```

One possible use-case would be if we add a `/timeseries/tiles` family of endpoints where you could easily get an xyz tile URL for a point along a timeseries. Maybe would be useful in a map application since you would have to load a single tile layer at a time. The `/timeseries/tilejson.json` endpoint gives you the whole set so maybe `step_idx`

I'm fine with leaving this parameter out but it might be useful someday.

Reply:
I actually wound up dropping the `step_idx` parameter in pursuit of a more succinct API.